src/shared/Microsoft.Git.CredentialManager/GitConfigurationKeyComparer.cs

```diff
@@ -0,0 +1,80 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT license.
+using System;
+
+namespace Microsoft.Git.CredentialManager
+{
+    /// <summary>
+    /// Represents string comparison of Git configuration entry key names.
+    /// </summary>
+    /// <remarks>
+    /// Git configuration entries have the form "section[.scope].property", where the
+    /// scope part is optional.
+    /// <para/>
+    /// The section and property components are NOT case sensitive.
+    /// The scope component if present IS case sensitive.
+    /// </remarks>
+    public class GitConfigurationKeyComparer : StringComparer
+    {
+        public static readonly GitConfigurationKeyComparer Instance = new GitConfigurationKeyComparer();
+
+        private GitConfigurationKeyComparer() { }
+
+        public override int Compare(string x, string y)
+        {
+            Split(x, out string xSection, out string xScope, out string xProperty);
+            Split(y, out string ySection, out string yScope, out string yProperty);
+
+            int cmpSection = OrdinalIgnoreCase.Compare(xSection, ySection);
+            if (cmpSection != 0) return cmpSection;
+
+            int cmpProperty = OrdinalIgnoreCase.Compare(xProperty, yProperty);
+            if (cmpProperty != 0) return cmpProperty;
+
+            return Ordinal.Compare(xScope, yScope);
+        }
+
+        public override bool Equals(string x, string y)
+        {
+            if (ReferenceEquals(x, y)) return true;
+            if (x is null || y is null) return false;
+
+            Split(x, out string xSection, out string xScope, out string xProperty);
+            Split(y, out string ySection, out string yScope, out string yProperty);
+
+            // Section and property names are not case sensitive, but the inner 'scope' IS case sensitive!
+            return OrdinalIgnoreCase.Equals(xSection, ySection) &&
+                   OrdinalIgnoreCase.Equals(xProperty, yProperty) &&
+                   Ordinal.Equals(xScope, yScope);
+        }
+
+        public override int GetHashCode(string obj)
+        {
+            Split(obj, out string section, out string scope, out string property);
+
+            int code = OrdinalIgnoreCase.GetHashCode(section) ^
+                       OrdinalIgnoreCase.GetHashCode(property);
+
+            return scope is null
+                ? code
+                : code ^ Ordinal.GetHashCode(scope);
+        }
+
+        private static void Split(string str, out string section, out string scope, out string property)
+        {
+            section = null;
+            scope = null;
+            property = null;
+
+            if (string.IsNullOrWhiteSpace(str))
+            {
+                return;
+            }
+
```

Comment:
Should we assert/throw/return If `str` doesn't have any dots in it?
Just wondering if the scopeLength computation goes negative.

Reply:
If the input string to either of `GetHashCode` or `Equals` (the only public methods) contains no dots, then we just behave like `StringComparer.Ordinal`. I'll add some tests to confirm this behaviour.